File: clojure/diamond/src/diamond.clj
(ns diamond (:require [clojure.string :as str]))

(def alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZ")

(defn- diamond-row [row-char col-heads]
  (apply str (for [col-char col-heads] (if (= row-char col-char) row-char \space))))

(defn diamond
  "Returns a diamond shape pattern for the given letter."
  [letter]
  (let [sub-alpha (subs alphabet 0 (inc (str/index-of alphabet letter)))
        rev-alpha (str/reverse sub-alpha)
        row-heads (str sub-alpha (subs rev-alpha 1))
        col-heads (str rev-alpha (subs sub-alpha 1))]
    (str/join "\n" (map #(diamond-row % col-heads) row-heads))))

File: python/diamond/diamond.py
"""Output a Diamond shape starting with A 
and having the supplied letter at its widest point"""

ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

def diamond_row(letter, columns):
    return ''.join(letter if letter == column else ' ' for column in columns)
        

def rows(letter):
    letters = ALPHABET[:ALPHABET.index(letter) + 1]
    row_letters = letters + letters[-2::-1]
    columns = letters[::-1] + letters[1:]

    return [diamond_row(row_letter, columns) for row_letter in row_letters]
